File: lib/domain/book.dart
/// Domain model for a Book in the user's personal library.
/// Maps directly to the BookResponseDto from the Spring Boot backend.
class Book {
  final int id;
  final String title;
  final String? author;
  final String? description;
  final String? isbn;
  final String? language;
  final String? category;
  final int? numberOfPages;
  final int? fileSize;
  final String? fileName;
  final bool isFavorite;
  final int rating;
  final DateTime createdAt;
  final DateTime updatedAt;

  Book({
    required this.id,
    required this.title,
    this.author,
    this.description,
    this.isbn,
    this.language,
    this.category,
    this.numberOfPages,
    this.fileSize,
    this.fileName,
    required this.isFavorite,
    required this.rating,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Book.fromJson(Map<String, dynamic> json) {
    return Book(
      id: json['id'],
      title: json['title'] ?? '',
      author: json['author'],
      description: json['description'],
      isbn: json['isbn'],
      language: json['language'],
      category: json['category'],
      numberOfPages: json['numberOfPages'],
      fileSize: json['fileSize'],
      fileName: json['fileName'],
      isFavorite: json['isFavorite'] ?? false,
      rating: json['rating'] ?? 0,
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }
}

File: eduvisionbackend/src/main/java/com/academicproject/eduvisionbackend/entity/Book.java
package com.academicproject.eduvisionbackend.entity;

import java.time.LocalDateTime;

import jakarta.persistence.*;
import lombok.*;

@Entity
@Table(name = "books")
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Builder
public class Book {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String title;

    @Column
    private String author;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column
    private String isbn;

    @Column
    private String language;

    @Column
    private String category;

    @Column
    private Integer numberOfPages;

    // Physical file storage path on server
    @Column(nullable = false, columnDefinition = "TEXT")
    private String filePath;

    @Column
    private Long fileSize;

    @Column
    private String fileName;

    @Column(name = "is_favorite", nullable = false)
    @Builder.Default
    private boolean isFavorite = false;

    @Column
    @Builder.Default
    private Integer rating = 0;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private User user;

    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}
